app/models/ability.rb

```diff
@@ -0,0 +1,31 @@
+# frozen_string_literal: true
+class Ability
+  include Hydra::Ability
+
+  def custom_permissions
+    can :manage, Valkyrie::Model if current_user.admin?
+    alias_action :file_manager, to: :update
+  end
+
+  def read_permissions
+    super
+    can :read, Valkyrie::Model do |obj|
+      valkyrie_test_read(obj) || valkyrie_test_edit(obj)
+    end
+  end
+
+  def edit_permissions
+    super
+    can [:edit, :update, :destroy], Valkyrie::Model do |obj|
+      valkyrie_test_edit(obj)
+    end
+  end
+
+  def valkyrie_test_read(obj)
+    !(user_groups & obj.read_groups).empty? || obj.read_users.include?(current_user.user_key)
```

Comment:
This is just copied from how `Hydra::AccessControls` does their checks. I think I'm cool with finding nicer patterns though, especially since the number of people who are likely to look at both are pretty slim.